src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs

```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
             return _getChildContentAsync(useCachedResult, encoder);
         }
 
-        /// <inheritdoc />
-        public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+        void IHtmlContentContainer.CopyTo(IHtmlContentBuilder destination)
+        {
+            _preElement?.CopyTo(destination);
+
+            var isTagNameNullOrWhitespace = string.IsNullOrWhiteSpace(TagName);
+
+            if (!isTagNameNullOrWhitespace)
+            {
+                destination.AppendHtml("<");
+                destination.AppendHtml(TagName);
+
+                // Perf: Avoid allocating enumerator
+                for (var i = 0; i < (_attributes?.Count ?? 0); i++)
+                {
+                    var attribute = _attributes[i];
+                    destination.AppendHtml(" ");
+                    destination.AppendHtml(attribute.Name);
+
+                    if (attribute.Minimized)
+                    {
+                        continue;
+                    }
+
+                    destination.AppendHtml("=\"");
+                    var value = attribute.Value;
+                    var htmlContent = value as IHtmlContent;
+                    if (htmlContent != null)
+                    {
+                        // Perf: static text in a bound attribute go down this path. Avoid allocating if possible (common case).
+                        var htmlEncodedString = value as HtmlEncodedString;
+                        if (htmlEncodedString != null && !htmlEncodedString.Value.Contains("\""))
+                        {
+                            destination.AppendHtml(htmlEncodedString);
+                        }
+                        else
+                        {
+                            destination.AppendHtml(new AttributeContent(htmlContent));
+                        }
+                    }
+                    else if (value != null)
+                    {
+                        destination.Append(value.ToString());
```

Comment:
Why `ToString()`? IOW what can `value` be other than a `string` or an `IHtmlContent` implementation?